### DISCUSSION TITLE
Hosted renovate private Azure Artifacts PyPi feed update support

### DISCUSSION THREAD

**Discussion by yogevyuval** [renovatebot/renovate#9614] [Request Help] [answered]
Hi,
Some of our python packages are private and are stored in a feed in Azure Artifacts. We use an --extra-index-url to fetch them as well with a token.

Is it possible to support updating these packages without self hosting renovate? How do we pass the token of the feed without committing them to source code? We have renovate running as a GitHub app.

From what I saw it's possible to encrypt values so only renovate would be able to read them, but the documentation is a bit unclear. First it looks likes it's only for npm? or is it just an example? Can we use it for any kind of token?

Also, how do we access the variable? is it turned into upper case from the original name in the renovate config?

Specifically im referring to:
https://docs.renovatebot.com/private-modules/

**Comment by rarkins** [comment, marked as answer]
1. Define a `hostRules` entry for the azure host
2. Use the hosted app's `encrypted` feature so that the token or password can be put into the config in encrypted form instead of raw

Something like this:

```
"hostRules": [{
  "hostName": "foo.some.azure.com",
  "encrypted": {
    "token": "asdgfsfasdfasfdasfasfasfdasfasfasfasfdasfasfa"
  }
}]
```

**Reply by yogevyuval** [reply]
@rarkins Thanks, So in your example the token will be available as ${TOKEN}?

**Reply by rarkins** [reply]
No, it will be converted into a regular hostRules entry with hostName and token. hostRules are then applied to each HTTP request generated by Renovate.

**Reply by yogevyuval** [reply]
Ok, sorry for being a bit repetitive, but I don't fully understand how to use this and the I can't find an example in the documentation.
Could you provide a more concrete example, or maybe provide a link to the correct page?

**Reply by rarkins** [reply]
You need too look up the documentation for `hostRules` and there are many examples all over this repo for those, particularly with tokens. Then look up the documentation for `encrypted`. And then finally here's the page where you encrypt: https://app.renovatebot.com/encrypt

**Reply by yogevyuval** [reply]
It finally worked.
For documentation purposes:
We had to use the AuthType: Basic, and decode the token to base64 before encrypting it as follows
base64(':token') - Note the ":" before the token, indicating an empty username

**Reply by viceice** [reply]
if you need basic auth, you should simply use `username` and `password` (pass token as password and empty username)

```json
"hostRules": [{
  "hostName": "foo.some.azure.com",
  "username": "",
  "encrypted": {
    "password": "asdgfsfasdfasfdasfasfasfdasfasfasfasfdasfasfa"
  }
}]
```

**Reply by yogevyuval** [reply]
I will try that, that's much more convenient, thanks!
The docs could really use a clarification around all of this.

**Reply by rarkins** [reply]
PRs are always welcome :). All docs are found in `docs/` in this repo

**Reply by YotamHershk** [reply]
> if you need basic auth, you should simply use `username` and `password` (pass token as password and empty username)
> 
> ```json
> "hostRules": [{
>   "hostName": "foo.some.azure.com",
>   "username": "",
>   "encrypted": {
>     "password": "asdgfsfasdfasfdasfasfasfdasfasfasfasfdasfasfa"
>   }
> }]
> ```

Hi, thanks for the suggested. That worked for us
